### PR TITLE
Detect collisions and generate parse errors when they are encountered

### PR DESCRIPTION
**Pull request by paulcam206**
Fix for #1852.

* source/shared/**
  * Introduce new `ParseContext.cpp` and `ParseContext.h` 
    * `ParseContext` encapsulates objects that need to be passed around at parse time
      * Add new `std::unordered_set<std::string>` to track ids
      * (previously: `ElementParserRegistration`, `ActionParserRegistration`, and `std::vector<AdaptiveParseWarning>`)
  * Various wiring to replace passing the old context items with `ParseContext`
  * New unit tests for single and nested duplicate id scenarios
  * Fixes to refrain from marking default dtors as virtual (cppcore)
  * Some `#include` cleanup
  * Reorder STL includes in `pch.h` for sanity's sake

* source/uwp/**
  * Update expected results for new failing tests
  * Add new files to shared model projection
  * Updates for new `Deserialize` signature

* samples/tests
  * tests for duplicate ids in single and nested scenarios

* source/android/**
  * Fixes to consume new header/object
  * SWIG'd changes

* source/dotnet/Library/adaptivecards/**
  * Introduce new `HashSet<string>` to track element ids
  * Add id tracking during [de]serialization (throw on collision)
  * Introduce a way to indicate when a card begins/ends parsing (needed to reset our id tracking member)
  * Account for expected failures for new tests

* source/ios/**
  * Publish/consume new cpp/h

### CODE REVIEW

File: source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/ParserRegistrationTest.cpp [hunk 1]
```diff
@@ -39,25 +39,21 @@ namespace AdaptiveCardsSharedModelUnitTest
...
-				virtual std::shared_ptr<BaseCardElement> Deserialize(
-					std::shared_ptr<AdaptiveCards::ElementParserRegistration> elementParserRegistration,
-					std::shared_ptr<AdaptiveCards::ActionParserRegistration> actionParserRegistration,
-					std::vector<std::shared_ptr<AdaptiveCards::AdaptiveCardParseWarning>>& warnings,
-					const Json::Value& value) override
+				virtual std::shared_ptr<BaseCardElement> Deserialize(ParseContext &context, const Json::Value& value) override
 				{
 					return std::make_shared<TestCustomElement>(value);
 				}
+                virtual std::shared_ptr<BaseCardElement> DeserializeFromString(ParseContext &context, const std::string& value) override
```
Comment (carlos-zamora) [inline source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/ParserRegistrationTest.cpp]: Indentation off?

File: source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/ResourceInformationTests.cpp [hunk 2]
```diff
@@ -230,25 +230,22 @@ namespace AdaptiveCardsSharedModelUnitTest
...
-                    std::shared_ptr<AdaptiveCards::ActionParserRegistration> actionParserRegistration,
-                    std::vector<std::shared_ptr<AdaptiveCards::AdaptiveCardParseWarning>>& warnings,
-                    const Json::Value& value) override
+
+                virtual std::shared_ptr<BaseCardElement> Deserialize(ParseContext &context, const Json::Value& value) override
                 {
                     return std::make_shared<TestCustomElement>(value);
                 }
+                virtual std::shared_ptr<BaseCardElement> DeserializeFromString(ParseContext &context, const std::string& value) override
+				{
```
Comment (carlos-zamora) [inline source/shared/cpp/AdaptiveCardsSharedModel/AdaptiveCardsSharedModelUnitTest/ResourceInformationTests.cpp]: (line 239-241) Indentation off?

File: source/shared/cpp/ObjectModel/ParseUtil.cpp [hunk 3]
```diff
@@ -310,20 +310,34 @@ namespace AdaptiveSharedNamespace
...
-    ParseUtil::GetElementCollection(std::shared_ptr<ElementParserRegistration> elementParserRegistration,
-                                    std::shared_ptr<ActionParserRegistration> actionParserRegistration,
-                                    std::vector<std::shared_ptr<AdaptiveCardParseWarning>>& warnings,
-                                    const Json::Value& json,
-                                    AdaptiveCardSchemaKey key,
-                                    bool isRequired)
+    template<typename T> static void AddId(const T& element, std::unordered_set<std::string>& ids)
+    {
+        auto id = element.GetId();
+        if (id != "")
```
Comment (RebeccaAnne) [inline source/shared/cpp/ObjectModel/ParseUtil.cpp]: >id != "" [](start = 12, length = 8)

I might do !id.empty()

File: source/shared/cpp/ObjectModel/SharedAdaptiveCard.cpp [hunk 4]
```diff
@@ -110,7 +91,32 @@ std::shared_ptr<ParseResult> AdaptiveCard::Deserialize(const Json::Value& json,
...
+
+    std::string version = ParseUtil::GetString(json, AdaptiveCardSchemaKey::Version, enforceVersion);
+    std::string fallbackText = ParseUtil::GetString(json, AdaptiveCardSchemaKey::FallbackText);
+    std::string language = ParseUtil::GetString(json, AdaptiveCardSchemaKey::Language);
+    std::string speak = ParseUtil::GetString(json, AdaptiveCardSchemaKey::Speak);
+
+    // check if language is valid
+    _ValidateLanguage(language, context.warnings);
+
+    // Perform version
```
Comment (RebeccaAnne) [inline source/shared/cpp/ObjectModel/SharedAdaptiveCard.cpp]: >Perform version [](start = 7, length = 15)

[nit] Enforce version? Perform version validation?

Comment (paulcam206) [reply source/shared/cpp/ObjectModel/SharedAdaptiveCard.cpp]: oops! :)

---
In reply to: [234325149](https://github.com/Microsoft/AdaptiveCards/pull/2171#discussion_r234325149) [](ancestors = 234325149)

**Pull request merged**